### PR TITLE
Ensembles

### PR DESCRIPTION
**Pull request by jindrahelcl**
Rebased version of #536 with included support for sequence decoder generalization

### CODE REVIEW

File: neuralmonkey/decoders/sequence_decoder.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    [("train_inputs", Optional[tf.Tensor])])
+
+DecoderFeedables = NamedTuple(
+    "DecoderFeedables",
+    [("step", tf.Tensor),  # 1D int, number of the step
+     ("finished", tf.Tensor),  # batch-sized, bool
+     ("input_symbol", tf.Tensor),
+     ("prev_logits", tf.Tensor)])
+
+LoopState = NamedTuple(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: add docs

File: neuralmonkey/decoders/sequence_decoder.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    "LoopState",
+    [("histories", Any),
+     ("constants", Any),
+     ("feedables", Any)])
+# pylint: enable=invalid-name
+
+
+# pylint: disable=too-many-public-methods
+# TODO More refactoring needed.. (?)
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: todo

File: neuralmonkey/decoders/sequence_decoder.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                 load_checkpoint: str = None) -> None:
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        log("Initializing decoder, name: '{}'".format(name))
+
+        self.vocabulary = vocabulary
+        self.data_id = data_id
+        self.max_output_len = max_output_len
+        self.dropout_keep_prob = dropout_keep_prob
+
+        # TODO check the values of the parameters (vocab, ...)
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: todo

File: neuralmonkey/decoders/sequence_decoder.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    def get_logits(self, state: tf.Tensor) -> tf.Tensor:
+        state = dropout(state, self.dropout_keep_prob, self.train_mode)
+        return tf.matmul(state, self.decoding_w) + self.decoding_b
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        # THE LAST TRAIN INPUT IS NOT USED IN DECODING FUNCTION
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: lowercase this

File: neuralmonkey/decoders/sequence_decoder.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        return tf.argmax(self.runtime_logits[:, :, 1:], -1) + 1
+
+    @tensor
+    def runtime_loss(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+        batch_major_logits = tf.transpose(self.runtime_logits, [1, 0, 2])
+        min_time = tf.minimum(tf.shape(train_targets)[1],
+                              tf.shape(batch_major_logits)[1])
+
+        # TODO if done properly, there should be padding of the shorter
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: what about this?

File: neuralmonkey/decoders/sequence_decoder.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            self.loop_continue_criterion,
+            self.get_body(train_mode, sample),
+            initial_loop_state)
+
+        self.finalize_loop(final_loop_state, train_mode)
+
+        logits = final_loop_state.histories.logits.stack()
+        decoder_outputs = final_loop_state.histories.decoder_outputs.stack()
+
+        # TODO mask should include also the end symbol
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_decoder.py]: well then.

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_decoder.py]: delete this (?)

File: neuralmonkey/decoders/sequence_decoder.py (hunk 5 above)
Comment (jlibovicky) [reply neuralmonkey/decoders/sequence_decoder.py]: create na issue for that

File: neuralmonkey/decoders/sequence_decoder.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    @tensor
+    def decoded(self) -> tf.Tensor:
+        return tf.argmax(self.runtime_logits[:, :, 1:], -1) + 1
+
+    @tensor
+    def runtime_loss(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+        batch_major_logits = tf.transpose(self.runtime_logits, [1, 0, 2])
+        min_time = tf.minimum(tf.shape(train_targets)[1],
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_decoder.py]: Neberes tu nahodou spatny rozmer? Po transpozici by train_targets mely mit shape [time, batch] a batch_major_logits [time, batch, vocab] a ty chces min_time (shape[0]), ne?

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_decoder.py]: Ne, u dekodérů je to opačně. Shape placeholderů je (time, batch). Chce to tam napsat na víc míst, pokud to tam nikde neni.

Je to kvůli tomu loopu. Když tam byl for-loop, tak se ty train_inputs rozsekaly podle první dimenze do pythonovýho pole v čase a pak se podle tohodle pole zase konkatenovaly.

Teď jsou zase tensorarraye taky v čase a když je stackneš, tak budeš mít čas v první dimenzi. Proto je to napsaný takhle debilně.

File: neuralmonkey/decoders/decoder.py [hunk 8]
```diff
@@ -398,15 +269,17 @@ def body(*args) -> LoopState:
                 # Run the RNN.
                 cell = self._get_rnn_cell()
                 if self._rnn_cell_str == 'GRU':
-                    cell_output, state = cell(rnn_input,
-                                              loop_state.prev_rnn_output)
+                    cell_output, state = cell(
+                        rnn_input, loop_state.feedables.prev_rnn_output)
```
Comment (varisd) [inline neuralmonkey/decoders/decoder.py]: Toto tady mame uz dlouhou, ale i tak se zeptam. Skutecne tam ma byt jako druhy argument prev_rnn_output (misto prev_rnn_state). Pokud jo, tak u GRU dekoderu vubec prev_rnn_state nevyuzivame.

Nebo jsem neco prehleld?

Comment (jlibovicky) [reply neuralmonkey/decoders/decoder.py]: V GRU je state a output to samé, ale je to tam odlišeně kvůli LSTM, která to má zvlášť (typicky značené jako _h_ a _c_).

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Ano, prev_rnn_state se v GRU dekoderu nepouziva. Nicméně to je potřeba feedovat když děláš beam search.. #550 by tohle mělo vyřešit

Comment (varisd) [reply neuralmonkey/decoders/decoder.py]: Ok, tak tady jde spis o terminologii. Preferoval bych predavat do GRU prev_rnn_state, vzhledem k tomu, ze i definice __call__(inputs, state, scope=None) to nazyva "state" (je to imho mene matouci).

Comment (jindrahelcl) [reply neuralmonkey/decoders/decoder.py]: Problém je hlavně v tom, že se slovo state používá na hrozně moc místech. Po #550 snad nebude tohle rozlišení vůbec potřeba.

File: neuralmonkey/decoders/beam_search_decoder.py [hunk 9]
```diff
@@ -222,87 +273,63 @@ def body(*args) -> BeamSearchLoopState:
...
+            next_loop_state = decoder_body(*dec_loop_state)  # type: ignore
 
-            next_prev_contexts = [tf.gather(ctx, next_beam_ids) for ctx in
-                                  next_loop_state.prev_contexts]
+            next_search_state = SearchState(
+                logprob_sum=next_beam_logprob_sum,
+                prev_logprobs=tf.nn.log_softmax(
+                    next_loop_state.feedables.prev_logits),
+                lengths=next_beam_lengths,
+                finished=dec_loop_state.feedables.finished)
```
Comment (varisd) [inline neuralmonkey/decoders/beam_search_decoder.py]: SearchState.finished nemuzes vytvaret na zaklade tf.gather(dec_loop_state.feedables.finished, next_beam_ids).

Vezmi napriklad situaci, ze v jednom kroku 2x picknes beam, ktery mel "<EOS>" jako symbol s nejvetsim logprobem (tim padem, je oznaceny jako finished). Jenze druhy nejpravdepodobnejsi symbol uz neni "<EOS>", takze ten by naopak nemel byt oznaceny jako finished.

Navic dec_loop_state.feedables.finished obsahuje hodnoty ktere vznikaji pred ensemblovanim.

Comment (jlibovicky) [reply neuralmonkey/decoders/beam_search_decoder.py]: Taky jsem si to myslel, ale pak jsem si řekl následující: ten seznam příznaků finished je ke každé hypotéze jeden, takže když tady volám `tf.gather` na finished podle toho, jaké vybírám hypotézy k pokračování, tak s každou hypotézou už pokračuje dál, jestli je finished a jedno, jestli, kolikátá byla v pořadí. Nebo ne?

Comment (varisd) [reply neuralmonkey/decoders/beam_search_decoder.py]: No jestli se nepletu, tak to tedka běží takto:
1. Spustí se decoder_body. To vygeneruje logity, vybere next_symbol (podle flagů: train, sample nebo argmax) a na základe next_symbol updatuje finished.

[Začátek iterace tf.while_loop]
(Zde je místo, kde se počítají logproby, které v runneru ensemblujeme a zpětně feedujeme.)

2. Na základě logprobů se spočítají beam_scores a vybere se top_k nových hypotéz. Některé tyto hypotézy ale mohou mít odlišný next_symbol než byl vybraný v 1. a je tedy přepočítat finished flagy pro tyto hypotézy.

3. S novýma hypotézama opět spouštíme decoder_body.
[Konec iterace tf.while_loop]

Myslím, že by se tam tedy měl vrátit to počítání "finished" pro top_k hypotézy.

Comment (jlibovicky) [reply neuralmonkey/decoders/beam_search_decoder.py]: Jo, ale ten gather tam nikde není koukám.

Comment (varisd) [reply neuralmonkey/decoders/beam_search_decoder.py]: https://github.com/ufal/neuralmonkey/blob/4db2be8fd6066b5b85d5bb2543a02c24197832cb/neuralmonkey/decoders/beam_search_decoder.py#L309

Poté, co se vytvoří next_feedables, tak by se mělo updatovat next_feedables.finished podle hodnot next_feedables.input_symbol.

**Comment by jlibovicky** [comment]
Bude se to dát merǧovat, až se vyřeší konflikt? Máme ověřeno na nečem větším, že to vážně funguje?

**Comment by varisd** [comment]
V neděli jsem dal trénovat překlad v aktuální větvi ensembles pro en-ja překlad. Do zítra bych mohl mít výsledky na testsetu.

Jinak jsem psal Jindrovi, ať mi dá natrénovat modely pro MMT, na kterých jsme to testovali naposledy. Tam ten trénink trvá podstatně kratší dobu.

**Comment by jindrahelcl** [comment]
Jakej je status tohohle? Modely samozřejmě nemam. Jak jsi na tom @varisd s ověřováním, že to funguje a nic to nekazí? (Přinejmenším naše obyčejný experimenty dopadaj pořád stejně, takže to vypadá v pohodě)

@jlibovicky  máš k tomuhle nějaký připomínky?

**Comment by jlibovicky** [comment]
Připomínku mám, že jsem to zkusil na OCR a ensembly snížily accuracy z 80% na 68%.

**Comment by varisd** [comment]
Dal jsem prekladat testsety k en-ja prekladu a na vsech datech ensembly konzistentne vylepsovaly BLEU ~1 bod. Pro jistotu jsem jeste meril ChrF-3 a vysledky byly taky pozitivni.

Na druhou stranu bych vytvoril issue tykajici se zrychleni ensemblu. Jedna veta se nyni preklada cca minutu, coz je jak z 90. let.

A jsem taky za to, aby se jeste zkusil ty modely z multimodal translation. Ty se netrenuji dlouho a dalsi potvrzeni by nebylo na skodu.

**Comment by jlibovicky** [comment]
Tak jsem se ujistil, že ty moje zhoršující výsledky jsou pravda, a co hůř, když dám ensemblovat 5× stejný model, tak to má taky horší výsledek.

**Comment by varisd** [comment]
To je divny (a hlavne blby). Posli mi mailem cestu k tomu configu a modelum. Pres tyden se na to zkusim podivat.

**Comment by jlibovicky** [comment]
Zkoušl jsi pustit ten můj kód, spravil jsi to? Víš, že ti neprochází testy?

**Comment by varisd** [comment]
Ten merge byl omyl. Jindra to uz opravil.
